core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java

```diff
@@ -154,6 +164,36 @@ public boolean binaryArithmeticCoercion(SqlCallBinding binding) {
     return coerced;
   }
 
+  /** For NUMERIC and DATE operands, cast NUMERIC to an INTERVAL_DAY.
+   *  <p>Only enable this coercion in Babel conformance mode.
+   */
+  protected boolean binaryArithmeticWithDateAndNumeric(
+      SqlCallBinding binding, RelDataType left, RelDataType right) {
+    // For expression "NUMERIC <OP> DATE" (or vice versa),
+    // TeradataSQL and Oracle would coerce the NUMERIC to an INTERVAL_DAY.
+    if (validator.getConformance() != SqlConformanceEnum.BABEL) {
+      return false;
+    }
+    int numericIndex;
+    if (SqlTypeUtil.isNumeric(left) && SqlTypeUtil.isDate(right)) {
+      numericIndex = 0;
+    } else if (SqlTypeUtil.isDate(left) && SqlTypeUtil.isNumeric(right)) {
+      numericIndex = 1;
+    } else {
+      return false;
+    }
+    SqlCall call = binding.getCall();
+    SqlNumericLiteral numeric = (SqlNumericLiteral) call.getOperandList().get(numericIndex);
+    SqlIntervalQualifier qualifier =
```

Comment:
How would you make sure the numeric operand is a literal ? Could it be another expression ?

Reply:
Why not just add a `CAST` there, let the planner decide how to translate to interval value.